code/datums/components/full_auto_fire.dm

```diff
@@ -0,0 +1,362 @@
+#define AUTOFIRE_MOUSEUP 0
+#define AUTOFIRE_MOUSEDOWN 1
+
+/datum/component/automatic_fire
+	var/client/clicker
+	var/mob/living/shooter
+	var/atom/target
+	var/autofire_stat = AUTOFIRE_STAT_SLEEPING
+	var/mouse_parameters
+	var/shots_fired = 0
+	var/autofire_shot_delay = 0.3 SECONDS //Time between individual shots.
+	var/auto_delay_timer
+	var/bursts_fired = 0
+	var/burstfire_shot_delay = 0.2 SECONDS
+	var/shots_to_fire = 0
+	var/component_fire_mode
+	var/mouse_status = AUTOFIRE_MOUSEUP //This seems hacky but there can be two MouseDown() without a MouseUp() in between if the user holds click and uses alt+tab, printscreen or similar.
+
+
+/datum/component/automatic_fire/Initialize(autofire_shot_delay, burstfire_shot_delay, shots_to_fire, firemode, parent_loc)
+	. = ..()
+	if(!isgun(parent))
+		return COMPONENT_INCOMPATIBLE
+	RegisterSignal(parent, COMSIG_GUN_FIREMODE_TOGGLE, .proc/wake_up)
+	RegisterSignal(parent, COMSIG_GUN_FIREDELAY_MODIFIED, .proc/modify_firedelay)
+	RegisterSignal(parent, COMSIG_GUN_BURSTDELAY_MODIFIED, .proc/modify_burst_delay)
+	RegisterSignal(parent, COMSIG_GUN_BURSTAMOUNT_MODIFIED, .proc/modify_burst_amount)
+
+	src.autofire_shot_delay = autofire_shot_delay
+	src.burstfire_shot_delay = burstfire_shot_delay
+	src.shots_to_fire = shots_to_fire
+
+	switch(firemode)
+		if(GUN_FIREMODE_AUTOMATIC, GUN_FIREMODE_AUTOBURST)
+			var/usercli
+			if(ismob(parent_loc))
+				var/mob/user = parent_loc
+				usercli = user.client
+			wake_up(src, firemode, usercli)
+
+
+/datum/component/automatic_fire/Destroy()
+	autofire_off()
+	return ..()
+
+
+/datum/component/automatic_fire/proc/wake_up(datum/source, fire_mode, client/usercli)
+	switch(fire_mode)
+		if(GUN_FIREMODE_SEMIAUTO, GUN_FIREMODE_BURSTFIRE)
+			switch(autofire_stat)
+				if(AUTOFIRE_STAT_IDLE, AUTOFIRE_STAT_ALERT, AUTOFIRE_STAT_FIRING)
+					sleep_up()
+			return //No need for autofire on this mode.
+	
+	component_fire_mode = fire_mode
+	
+	switch(autofire_stat)
+		if(AUTOFIRE_STAT_IDLE, AUTOFIRE_STAT_ALERT)

```

Comment:
In here there are 4 options and 3 possible behaviors:
If 1 or 2, return.
If 3, stop autofiring and return.
If 4 go on.

Would require two `if` checks to substitute this, and I feel a `switch` is better.

Reply:
Why is it better? Yes you get slight performance benefits with a switch but in most cases `if` statements are more readable and don't have as much indentation. It's fine if it's being used in this case but I'd really consider not doing it regardless because it requires more time invested in reading the code to understand exactly what you want.